### ISSUE TITLE
$(RELEASE_PATH) need some modify: distillery 1.0.0

### ISSUE THREAD

**Issue by luckynum7** [edib-tool/mix-edib#28]
Hi, I run the default `mix edib` with distillery 1.0.0 , and get the following error:

```bash
| make -f edib/make-tarball.mk
| make[2]: Entering directory '/build'
| Packaging your app ...
| mkdir -p /stage/rootfs
| cp -vfa --parents /bin/busybox /stage/rootfs/
| '/bin/busybox' -> '/stage/rootfs/bin/busybox'
| /bin/busybox --install -s /stage/rootfs/bin
| mkdir -p /stage/rootfs/app
| tar -xzf /build/app/rel/chatty/releases/0.0.1/chatty.tar.gz -C /stage/rootfs/app
| tar: can't open '/build/app/rel/chatty/releases/0.0.1/chatty.tar.gz': No such file or directory
| make[2]: *** [edib/make-tarball.mk:61: /stage/rootfs/app/bin] Error 1
| make[2]: Leaving directory '/build'
| make[1]: Leaving directory '/build'
| make[1]: *** [edib/edib.mk:20: /stage/tarballs/20161218_154215/chatty-0.0.1.tar.gz] Error 2
| make: *** [edib/Makefile:6: edib] Error 2
```

The issue https://github.com/edib-tool/docker-edib-tool/issues/6 inspired me, so I rebuild [docker-elixir-dev](https://github.com/edib-tool/docker-elixir-dev) (to update the hex version to 0.14.1) and [docker-edib-tool](https://github.com/edib-tool/docker-edib-tool). And modify the edib/shared.mk:

```diff
diff --git a/edib/shared.mk b/edib/shared.mk
index 1924937..694fd58 100644
--- a/edib/shared.mk
+++ b/edib/shared.mk
@@ -10,7 +10,7 @@ APP_VER        = $(shell $(APPINFO_RUNNER) version)
 
 MIX_ENV       ?= prod
 RELEASE        = releases/$(APP_VER)/$(APP_NAME).tar.gz
-RELEASE_PATH   = $(APP_DIR)/rel/$(APP_NAME)
+RELEASE_PATH   = $(APP_DIR)/_build/$(MIX_ENV)/rel/$(APP_NAME)
 RELEASE_FILE   = $(RELEASE_PATH)/$(RELEASE)
 
 STAGE_DIR      = /stage
```

As a result, the build successed:

```bash
==> Packaging was successful! \o/
```

*******

However, the command:

```bash
docker run --rm -e "PORT=4000" -p 4000:4000 local/chatty:0.0.1
```

show the following error:

```bash
standard_init_linux.go:175: exec user process caused "no such file or directory"
```

Any suggestion? Thanks.

**Comment by luckynum7** [comment]
I checked the generated `chatty-0.0.1.tar.gz` with

```bash
$ tar -ztvf server/tarballs/chatty-0.0.1.tar.gz|rg '/bin/'
3:drwxr-xr-x root/root         0 2016-12-19 08:46 ./app/bin/
4:-rwxrwxrwx root/root       750 2016-12-19 08:46 ./app/bin/chatty
5:-rw-r--r-- root/root      5901 2016-12-19 08:46 ./app/bin/nodetool
6:-rw-r--r-- root/root     19060 2016-12-19 08:46 ./app/bin/start_clean.boot
7:-rw-r--r-- root/root     10835 2016-12-19 08:46 ./app/bin/release_utils.escript
889:drwxr-xr-x root/root         0 2016-12-19 08:46 ./app/erts-8.0.1/bin/
...
910:-rwxr-xr-x root/root      9912 2016-12-19 08:46 ./app/erts-8.0.1/bin/heart
1013:drwxr-xr-x root/root         0 2016-12-19 08:46 ./bin/
...
1278:-rwxr-xr-x root/root    821408 2016-09-26 21:45 ./bin/busybox
1279:lrwxrwxrwx root/root         0 2016-12-19 08:46 ./bin/gunzip -> /bin/busybox
...
1317:lrwxrwxrwx root/root         0 2016-12-19 08:46 ./bin/fstrim -> /bin/busybox
```
full output: https://gist.github.com/luckynum7/fe5f3536c0fe0fdf708031ee5539fb78#file-tarball-bin-md

However, both `/bin/busybox` and `/app/bin/chatty` give the same error:

```bash
$ docker run -it --rm --entrypoint=/bin/busybox local/chatty:0.0.1
standard_init_linux.go:175: exec user process caused "no such file or directory"
```

```bash
$ docker run -it --rm --entrypoint=/app/bin/chatty local/chatty:0.0.1
standard_init_linux.go:175: exec user process caused "no such file or directory"
````

**Comment by luckynum7** [comment]
Hi, I use `docker save` to make the generated image into a tarball.

```bash
$ docker save local/chatty:0.0.1 -o tar/chatty.tar
```

then `tar xf` and `tar xf` to list the files

```bash
$ ls
3024bf322efbf8c364111da8e513fb5071470665b16c0fa572edbe81f42624fe.json  chatty.tar     repositories
b9372c49ca0e817c66c195f5fb92047e83a67ab14ef5345e1f2a70ef7b0a5cea       manifest.json
$ cd b9372c49ca0e817c66c195f5fb92047e83a67ab14ef5345e1f2a70ef7b0a5cea 
$ ls
app  bin  json  layer.tar  VERSION
```

```bash
$ app/bin/chatty 
Usage: chatty {start|start_boot <file>|foreground|stop|restart|reboot|pid|ping|console|console_clean|console_boot <file>|attach|remote_console|upgrade|downgrade|install|escript|rpc|rpcterms|eval|command <module> <function> <args>}
$ app/bin/chatty console
/home/hoffmann/Project/elixir-is-fun/tar/b9372c49ca0e817c66c195f5fb92047e83a67ab14ef5345e1f2a70ef7b0a5cea/app/releases/0.0.1/chatty.sh: line 614: /home/hoffmann/Project/elixir-is-fun/tar/b9372c49ca0e817c66c195f5fb92047e83a67ab14ef5345e1f2a70ef7b0a5cea/app/erts-8.0.1/bin/erlexec: No such file or directory
$ app/erts-8.0.1/bin/erlexec 
zsh: no such file or directory: app/erts-8.0.1/bin/erlexec
$ file app/erts-8.0.1/bin/erlexec 
app/erts-8.0.1/bin/erlexec: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-musl-x86_64.so.1, stripped
```

Any idea?

**Comment by luckynum7** [comment]
Btw, `/bin/busybox` works in the docker image `alpine:3.4`.

```bash
$ docker run -it --rm --entrypoint=/bin/busybox alpine:3.4
BusyBox v1.24.2 (2016-08-12 14:38:34 GMT) multi-call binary.
BusyBox is copyrighted by many authors between 1998-2015.
Licensed under GPLv2. See source distribution for detailed
copyright notices.

...
```

full output: https://gist.github.com/luckynum7/fe5f3536c0fe0fdf708031ee5539fb78#file-busybox-alpine-3-4-md

**Comment by luckynum7** [comment]
Update, older title for the issue:
$(RELEASE_PATH) may need some modify: "tar: ... No such file or directory"

According to the changelog for `distillery`: https://github.com/bitwalker/distillery/blob/master/CHANGELOG.md#100

```
Default output path for release artifacts is _build/<env>/rel/<relname>, 
this can be configured in rel/config.exs with set output_dir: "path"
```

`mix edib` full output:
https://gist.github.com/luckynum7/fe5f3536c0fe0fdf708031ee5539fb78#file-mix-edib-md

**Comment by 0xCAmM3r** [comment]
@luckynum7 I'm getting this same error, and a different error with 0.10 distillery. I was just about to go down the same path you were, so I'm glad I saw this, and I'm looking forward to yer commit!

**Comment by rowpsmo1859** [comment]
@luckynum7 Thank you for the wonderful detailed issue. Using the contents of your changes. I was successful at building a docker image that runs a phoenix server, using `distillery 1.1.0`

I'll push up the commits and create some PRs

**Comment by rowpsmo1859** [comment]
We'll need more testing on this to make sure this doesn't break others.

I had an issue with a git dependency in my project but was able to switch to a hex package as a workaround. Should be investigated further.

**Comment by rowpsmo1859** [comment]
Cancel that. Ended up running an old docker image on accident. Troubleshooting

EDIT: Was able to generate and run a packaged release using distillery. Must be something in the docker image then.

After the docker image crashes run `journalctl -u docker.service`

**Comment by rowpsmo1859** [comment]
Here's what I've found out so far

The generated docker image is unable to be started using the following command
`docker run -it generated-docker-image/name /bin/sh`

-------------

Searching for answers has revealed that the error message
```
panic: standard_init_linux.go:178: exec user process caused "no such file or directory" [recovered]
	panic: standard_init_linux.go:178: exec user process caused "no such file or directory"
```

is potentially caused by a line ending issue https://github.com/docker/docker/issues/27940

Not sure how that fits in yet. Either way the error message isn't too helpful.

------------

Comparing an old generated docker image to a new one didn't reveal anything

To do so run
`docker inspect generated-docker-image/name`
on two images then compare the output

----------------

## What needs to be done

Not being able to launch the docker image in a shell instance is the big issue here. I'm not certain what is causing it yet but I have some ideas for experiments to narrow the possible offending code.

@ryancammer says that this issue is occurring in distillery version 0.10 (the version before 1.0). Based on that, the issue may not be caused by distillery. It may be a recent change in mix-edib or one of the docker images in the docker build chain. (Should run an experiment to solidify this)

I'm going to take a break and absorb what I've learned thus far. Hopefully that will reveal a path forward. 

--------------------

Any ideas?

**Comment by luckynum7** [comment]
@JeremyBellows, I fired a PR couple of days ago, https://github.com/edib-tool/docker-edib-tool/pull/9 and forgot to update here. To build the correct image, there's another change need to be made

```diff
diff --git a/tools/libdeps.exs b/tools/libdeps.exs
index 2399b70..4b93518 100755
--- a/tools/libdeps.exs
+++ b/tools/libdeps.exs
@@ -1,6 +1,6 @@
 *!/usr/bin/env elixir
 defmodule Libdeps do
-  @relpath "app/rel"
+  @relpath "app/_build"
   @lddpath_regex ~r/\/(lib|usr\/lib)[^ ]+/

   def all_files do
```

However,  this made the `relpath` different from the original one. The reason is that I don't know how to pass the MIX.ENV down to it. The related files: `tools/libdeps`, `edib/make-tarball.mk`.

**Comment by asaaki** [comment]
Just change it to:

```diff
 #!/usr/bin/env elixir
 defmodule Libdeps do
-  @relpath "app/rel"
+  @current_mix_env System.get_env("MIX_ENV") || "prod"
+  @relpath "app/_build/#{@current_mix_env}"
```

Also see comment on <https://github.com/edib-tool/docker-edib-tool/pull/10>.

**Comment by rowpsmo1859** [comment]
@luckynum7 Excellent! I should've checked the existing PRs. My bad!
Would you mind sharing your process for solving this issue?

@asaaki I closed one of the PRs I opened in favor of @luckynum7 PR. Let me know if there is anything else I can do :)

Great job guys!

**Comment by asaaki** [comment]
@luckynum7 @ryancammer @JeremyBellows Could you check again? The image chain was rebuilt recently multiple times, lots of dependencies should have been updated as well. Might be that the described issues have dissolved now.

**Comment by luckynum7** [comment]
@asaaki I tried the latest commit of edib-tool/mix-edib, same error:

```bash
| make -f edib/make-tarball.mk
| make[2]: Entering directory '/build'
| Packaging your app ...
| mkdir -p /stage/rootfs
| cp -vfa --parents /bin/busybox /stage/rootfs/
| '/bin/busybox' -> '/stage/rootfs/bin/busybox'
| /bin/busybox --install -s /stage/rootfs/bin
| mkdir -p /stage/rootfs/app
| tar -xzf /build/app/rel/chatty/releases/0.0.1/chatty.tar.gz -C /stage/rootfs/app
| tar: can't open '/build/app/rel/chatty/releases/0.0.1/chatty.tar.gz': No such file or directory
| make[2]: Leaving directory '/build'
| make[2]: *** [edib/make-tarball.mk:61: /stage/rootfs/app/bin] Error 1
| make[1]: *** [edib/edib.mk:20: /stage/tarballs/20170122_005702/chatty-0.0.1.tar.gz] Error 2
| make[1]: Leaving directory '/build'
| make: *** [edib/Makefile:6: edib] Error 2
==> An error happened!
==> Reason: Artifact tarball not created! Check for any errors above.
```

full output: https://gist.github.com/luckynum7/fe5f3536c0fe0fdf708031ee5539fb78#file-edib-tool-1-5-2-md

**Comment by luckynum7** [comment]
@asaaki I update the `relpath` to respect the `MIX_ENV` based on your comment, thanks.
Updated PR: https://github.com/edib-tool/docker-edib-tool/pull/9

@JeremyBellows The process for solving this issue: It took me some time to chase down the output as well as those rebuild docker images, and finally found the `libdeps` file and the `relpath`.

**Comment by asaaki** [comment]
I have released [mix-edib v0.10.0](https://github.com/edib-tool/mix-edib/releases/tag/v0.10.0) which uses [edib-tool image v1.6.0](https://github.com/edib-tool/docker-edib-tool/releases/tag/1.6.0).

_Btw: You can also specify the docker image via `mix edib --edib edib/edib-tool:1.6.0` now. Useful if you want to test different image versions. (Maybe the option should be renamed to `--tool` in the future.)_ 😅
